File: app/src/main/java/com/bortolan/iquadriv2/Fragments/Orario.java
package com.bortolan.iquadriv2.Fragments;


import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.bortolan.iquadriv2.Adapters.AdapterOrari;
import com.bortolan.iquadriv2.Databases.FavouritesDB;
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubItem;
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubResponse;
import com.bortolan.iquadriv2.R;
import com.bortolan.iquadriv2.Tasks.CacheObjectObservable;
import com.bortolan.iquadriv2.Tasks.CacheObjectTask;
import com.bortolan.iquadriv2.Utils.DownloadSchedules;
import com.bortolan.iquadriv2.Views.ScheduleList;

import java.io.File;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.schedulers.Schedulers;

import static com.bortolan.iquadriv2.Utils.Methods.isNetworkAvailable;

public class Orario extends Fragment implements AdapterOrari.UpdateFragment, SearchView.OnQueryTextListener {
    private final static String TAG = Orario.class.getSimpleName();
    Context mContext;
    @BindView(R.id.search_card)
    CardView searchCard;
    @BindView(R.id.search_view)
    SearchView searchView;
    @BindView(R.id.recycler)
    RecyclerView recyclerView;
    @BindView(R.id.preferiti)
    ScheduleList preferiti;
    @BindView(R.id.classi)
    ScheduleList classi;
    @BindView(R.id.prof)
    ScheduleList prof;
    @BindView(R.id.aule)
    ScheduleList aule;

    FavouritesDB db;

    public Orario() {
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_orario, container, false);
        ButterKnife.bind(this, layout);
        mContext = getContext();
        db = FavouritesDB.getInstance(mContext);

        searchView.setOnQueryTextListener(this);

        bindSchedulesCache();
        if (isNetworkAvailable(mContext)) {
            new DownloadSchedules(response -> addAll(response, true)).execute();
        }
        return layout;
    }

    @Override
    public void onResume() {
        super.onResume();

        preferiti.setData("preferiti", db.getAll(), db, this, false);
    }

    public void addAll(GitHubResponse response, boolean docache) {
        if (response != null) {
            classi.setData("classi", response.getClassi(), db, this, true);
            prof.setData("prof", response.getProf(), db, this, true);
            aule.setData("aule", response.getAule(), db, this, true);

            if (docache) {
                new CacheObjectTask(mContext.getCacheDir(), TAG).execute(response);
            }
        }
    }

    public void bindSchedulesCache() {
        new CacheObjectObservable(new File(mContext.getCacheDir(), TAG)).getCachedObject(GitHubResponse.class)
                .subscribeOn(Schedulers.io())
                .subscribe(response -> {
                    classi.setData("classi", response.getClassi(), db, this, true);
                    prof.setData("prof", response.getProf(), db, this, true);
                    aule.setData("aule", response.getAule(), db, this, true);
                }, Throwable::printStackTrace);
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        preferiti.filter(newText);
        classi.filter(newText);
        prof.filter(newText);
        aule.filter(newText);
        return true;
    }

    @Override
    public void add(GitHubItem item, int position) {
        preferiti.add(item, position);
    }

    @Override
    public void remove(GitHubItem item, int position) {
        preferiti.remove(position);
    }

    @Override
    public void onStop() {
        super.onStop();
        FavouritesDB.getInstance(getContext()).close();
    }
}

File: app/src/main/java/com/bortolan/iquadriv2/Fragments/Orario.kt
package com.bortolan.iquadriv2.Fragments


import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.SearchView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.bortolan.iquadriv2.Adapters.AdapterOrari
import com.bortolan.iquadriv2.Databases.FavouritesDB
import com.bortolan.iquadriv2.Databases.RegistroDB
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubItem
import com.bortolan.iquadriv2.Interfaces.GitHub.GitHubResponse
import com.bortolan.iquadriv2.R
import com.bortolan.iquadriv2.Tasks.Remote.DownloadSchedules
import com.bortolan.iquadriv2.Utils.Methods
import com.bortolan.iquadriv2.Widget.WidgetOrario
import kotlinx.android.synthetic.main.fragment_orario.*


class Orario : Fragment(), AdapterOrari.UpdateFragment, SearchView.OnQueryTextListener {
    var active = false
    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_orario, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        search_view.setOnQueryTextListener(this)
        search_view.findViewById<ImageView>(R.id.search_close_btn).setOnClickListener {
            search_view.clearFocus()
            preferiti.requestFocus()
            search_view.setQuery(null, true)
            preferiti.filter(null)
        }
    }

    override fun onResume() {
        super.onResume()

        if (!search_view.query.isNotEmpty()) {
            preferiti.setData("preferiti", FavouritesDB.getInstance(context).all, this, false)
            load()
            download()
        }

        active = true
    }

    override fun onStop() {
        super.onStop()
        active = false
        context.sendBroadcast(Intent(context, WidgetOrario::class.java).setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE).putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS,
                AppWidgetManager.getInstance(context).getAppWidgetIds(ComponentName(context, WidgetOrario::class.java))))
    }

    fun addAll(response: GitHubResponse) {
        classi?.setData("classi", response.classi, this, true)
        prof?.setData("professori", response.prof, this, true)
        aule?.setData("aule", response.aule, this, true)
    }

    fun download() {
        val mContext: Context = context
        if (Methods.isNetworkAvailable(mContext)) {
            DownloadSchedules { response: GitHubResponse ->
                RegistroDB.getInstance(mContext).addSchedules(response)
                if (active) addAll(RegistroDB.getInstance(mContext).schedules)
            }.execute()
        }
    }

    fun load() {
        addAll(RegistroDB.getInstance(context).schedules)
    }

    override fun onQueryTextSubmit(query: String?): Boolean {
        search_view.clearFocus()
        return true
    }

    override fun onQueryTextChange(newText: String): Boolean {
        preferiti.filter(newText)
        classi.filter(newText)
        prof.filter(newText)
        aule.filter(newText)
        return true
    }

    override fun add(item: GitHubItem, position: Int) {
        preferiti.add(item)
    }

    override fun remove(item: GitHubItem, position: Int) {
        preferiti.remove(position)
    }

    companion object {
        private val TAG = "Orario"
    }
}
